File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/SystemUINetworkSpeeddoublelayerHook.java
package com.qimian233.ztool.hook.modules.systemui;

import android.annotation.SuppressLint;
import android.text.Html;

import com.qimian233.ztool.hook.base.BaseHookModule;
import com.qimian233.ztool.hook.base.DexKitHelper;

import io.github.libxposed.api.XposedModuleInterface;

import org.luckypray.dexkit.DexKitBridge;
import org.luckypray.dexkit.query.FindClass;
import org.luckypray.dexkit.query.matchers.ClassMatcher;
import org.luckypray.dexkit.result.ClassData;

import java.lang.reflect.Constructor;
import java.lang.reflect.Method;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.WeakHashMap;

/**
 * SystemUI网络速度显示Hook模块
 * 功能：在状态栏显示实时上下行网络速度，支持自定义文本大小和显示格式
 */
@SuppressLint("PrivateApi")
public class SystemUINetworkSpeeddoublelayerHook extends BaseHookModule {

    private static final String SYSTEMUI_PACKAGE = "com.android.systemui";
    private static final String NETWORK_SPEED_VIEW_CLASS = "com.android.systemui.zui.NetworkSpeedView";

    // 存储每个实例的上次流量数据
    private final Map<Object, Long> lastRxBytesMap = new WeakHashMap<>();
    private final Map<Object, Long> lastTxBytesMap = new WeakHashMap<>();
    private final Map<Object, Long> lastUpdateTimeMap = new WeakHashMap<>();

    public SystemUINetworkSpeeddoublelayerHook() {}

    @Override
    public String getModuleName() {
        return "systemui_network_speed_doublelayer";
    }

    @Override
    public String[] getTargetPackages() {
        return new String[]{SYSTEMUI_PACKAGE};
    }

    @Override
    public void handleLoadPackage(XposedModuleInterface.PackageLoadedParam param) throws Throwable {
        ClassLoader classLoader = param.getDefaultClassLoader();
        String packageName = param.getPackageName();
        if (SYSTEMUI_PACKAGE.equals(packageName)) {
            hookSystemUINetworkSpeed(classLoader);
        }
    }

    private void hookSystemUINetworkSpeed(ClassLoader classLoader) {
        try {
            log("Starting to hook SystemUI NetworkSpeedView");

            // Hook NetworkSpeedView 构造方法
            Constructor<?> ctor = classLoader.loadClass(NETWORK_SPEED_VIEW_CLASS)
                    .getDeclaredConstructor(android.content.Context.class,
                            android.util.AttributeSet.class,
                            int.class);
            this.xposed.hook(ctor).intercept(chain -> {
                chain.proceed();
                initNetworkSpeedView(chain.getThisObject());
                return null;
            });

            // Hook Handler 的 handleMessage 方法
            hookNetworkSpeedHandler(classLoader);

            log("Successfully hooked NetworkSpeedView");

        } catch (Throwable t) {
            logError("Error hooking NetworkSpeedView", t);
        }
    }

    private void initNetworkSpeedView(Object networkSpeedView) {
        try {
            Class<?> cl = networkSpeedView.getClass();
            // 获取初始流量数据
            long initialRxBytes = getTotalRxBytes();
            long initialTxBytes = getTotalTxBytes();

            // 存储初始数据
            lastRxBytesMap.put(networkSpeedView, initialRxBytes);
            lastTxBytesMap.put(networkSpeedView, initialTxBytes);
            lastUpdateTimeMap.put(networkSpeedView, System.currentTimeMillis());

            // 调整文本大小
            try {
                // 获取当前文本大小并增加
                Method getTextSizeMethod = findMethod(cl, "getTextSize");
                Object textSizeResult = getTextSizeMethod.invoke(networkSpeedView);
                float currentTextSize = textSizeResult != null ? (Float) textSizeResult : 8.0f;
                float newTextSize = currentTextSize * 1.1f; // 增加10%

                Method setTextSizeMethod = findMethod(cl, "setTextSize", int.class, float.class);
                setTextSizeMethod.invoke(
                        networkSpeedView, android.util.TypedValue.COMPLEX_UNIT_PX, newTextSize);

                if (DEBUG) {
                    log("Adjusted text size from " + currentTextSize + " to " + newTextSize);
                }
            } catch (Throwable sizeError) {
                logError("Error adjusting text size", sizeError);
            }

            log("Initialized NetworkSpeedView instance");
        } catch (Throwable t) {
            logError("Error initializing NetworkSpeedView", t);
        }
    }

    private void hookNetworkSpeedHandler(ClassLoader classLoader) {
        try {
            // 通过 DEXKit 查找 NetworkSpeedView 的内部 Handler 类（替代硬编码 $3）
            Class<?> handlerClass = findHandlerInnerClass(classLoader);

            Method handleMessageMethod = handlerClass.getDeclaredMethod("handleMessage", android.os.Message.class);
            this.xposed.hook(handleMessageMethod).intercept(chain -> {
                Object handler = chain.getThisObject();
                Class<?> handlerCls = handler.getClass();
                java.lang.reflect.Field this0Field = handlerCls.getDeclaredField("this$0");
                this0Field.setAccessible(true);
                Object networkSpeedView = this0Field.get(handler);

                // 获取消息对象
                Object message = chain.getArg(0);
                int what = message.getClass().getDeclaredField("what").getInt(message);

                if (what == 10) { // 更新速度的消息
                    handleSpeedUpdate(networkSpeedView, handler);
                    return null; // 阻止原始处理
                } else if (what == 1) { // 格式化显示的消息
                    handleSpeedDisplay(networkSpeedView, message);
                    return null; // 阻止原始处理
                }
                return chain.proceed();
            });

        } catch (Throwable t) {
            logError("Error hooking NetworkSpeed handler", t);
        }
    }

    private void handleSpeedUpdate(Object networkSpeedView, Object handler) {
        try {
            Class<?> handlerCls = handler.getClass();
            // 移除之前的消息
            findMethod(handlerCls, "removeMessages", int.class).invoke(handler, 10);

            // 检查是否应该显示网速
            Object isIconVisibleResult = networkSpeedView.getClass()
                    .getDeclaredMethod("isIconVisible").invoke(networkSpeedView);
            boolean shouldShow = Boolean.TRUE.equals(isIconVisibleResult);

            if (!shouldShow) {
                return;
            }

            // 获取当前流量统计
            long currentRxBytes = getTotalRxBytes();
            long currentTxBytes = getTotalTxBytes();
            long currentTime = System.currentTimeMillis();

            // 获取上次的数据
            Long lastRxBytes = lastRxBytesMap.get(networkSpeedView);
            Long lastTxBytes = lastTxBytesMap.get(networkSpeedView);
            Long lastUpdateTime = lastUpdateTimeMap.get(networkSpeedView);

            if (lastRxBytes == null || lastTxBytes == null || lastUpdateTime == null) {
                // 第一次更新，只记录数据不计算速度
                lastRxBytesMap.put(networkSpeedView, currentRxBytes);
                lastTxBytesMap.put(networkSpeedView, currentTxBytes);
                lastUpdateTimeMap.put(networkSpeedView, currentTime);
            } else {
                // 计算时间差（秒）
                long timeDiff = (currentTime - lastUpdateTime) / 1000;
                if (timeDiff > 0) {
                    // 计算上下行速度（字节/秒）
                    long downSpeed = (currentRxBytes - lastRxBytes) / timeDiff;
                    long upSpeed = (currentTxBytes - lastTxBytes) / timeDiff;

                    // 记录调试信息
                    if (DEBUG) {
                        log(String.format(Locale.US, "Successfully updated speed - downSpeed=%d, upSpeed=%d, timeDiff=%d",
                                downSpeed, upSpeed, timeDiff));
                    }

                    // 发送显示消息
                    Object message = findMethod(handlerCls, "obtainMessage").invoke(handler);
                    if (message != null) {
                        Class<?> msgCls = message.getClass();
                        msgCls.getDeclaredField("what").setInt(message, 1);
                        msgCls.getDeclaredField("obj").set(message, new long[]{downSpeed, upSpeed});
                        findMethod(handlerCls, "sendMessage", android.os.Message.class)
                                .invoke(handler, message);
                    }

                    // 更新数据
                    lastRxBytesMap.put(networkSpeedView, currentRxBytes);
                    lastTxBytesMap.put(networkSpeedView, currentTxBytes);
                    lastUpdateTimeMap.put(networkSpeedView, currentTime);
                }
            }

            // 安排下一次更新（3秒后）
            findMethod(handlerCls, "sendEmptyMessageDelayed", int.class, long.class)
                    .invoke(handler, 10, 3000L);

        } catch (Throwable t) {
            logError("Error in speed update", t);
        }
    }

    private void handleSpeedDisplay(Object networkSpeedView, Object message) {
        try {
            long[] speeds = (long[]) message.getClass().getDeclaredField("obj").get(message);
            if (speeds == null || speeds.length != 2) return;

            long downSpeed = speeds[0];
            long upSpeed = speeds[1];

            // 格式化上下行速度
            String downText = formatSpeed(downSpeed);
            String upText = formatSpeed(upSpeed);

            // 创建带有HTML格式的双层显示文本，调整字体大小
            String displayText = "<font size='5'><b>▴ " + upText + "</b></font><br/>" +
                    "<font size='5'><b>▾ " + downText + "</b></font>";

            // 使用HTML格式设置文本
            findMethod(networkSpeedView.getClass(), "setText", CharSequence.class)
                    .invoke(networkSpeedView,
                            android.text.Html.fromHtml(displayText, Html.FROM_HTML_MODE_LEGACY));

        } catch (Throwable t) {
            logError("Error in speed display", t);
        }
    }

    private String formatSpeed(long speedBytes) {
        if (speedBytes <= 0) {
            return "0.00 B/s";
        }

        double speed;
        String unit;

        if (speedBytes >= 1073741824) { // 1 GB
            speed = speedBytes / 1073741824.0;
            unit = "G/s";
        } else if (speedBytes >= 1048576) { // 1 MB
            speed = speedBytes / 1048576.0;
            unit = "M/s";
        } else if (speedBytes >= 1024) { // 1 KB
            speed = speedBytes / 1024.0;
            unit = "K/s";
        } else {
            speed = speedBytes;
            unit = "B/s";
        }

        // 根据速度值选择合适的精度
        String formatPattern;
        if (speed >= 100) {
            formatPattern = "0";
        } else if (speed >= 10) {
            formatPattern = "0.0";
        } else {
            formatPattern = "0.00";
        }

        java.text.DecimalFormat df = new java.text.DecimalFormat(formatPattern);
        return df.format(speed) + " " + unit;
    }

    private long getTotalRxBytes() {
        try {
            Object result = android.net.TrafficStats.class
                    .getDeclaredMethod("getTotalRxBytes").invoke(null);
            return result != null ? (Long) result : 0L;
        } catch (Throwable t) {
            logError("Error getting Rx bytes", t);
            return 0;
        }
    }

    private long getTotalTxBytes() {
        try {
            Object result = android.net.TrafficStats.class
                    .getDeclaredMethod("getTotalTxBytes").invoke(null);
            return result != null ? (Long) result : 0L;
        } catch (Throwable t) {
            logError("Error getting Tx bytes", t);
            return 0;
        }
    }

    /**
     * 通过反射查找 NetworkSpeedView 的内部 Handler 子类。
     * 遍历可能的内部类索引，替代硬编码的 $3。
     */
    private Class<?> findHandlerInnerClass(ClassLoader classLoader) {
        // 先尝试 DEXKit 查找 Handler 子类
        DexKitBridge bridge = DexKitHelper.INSTANCE.getBridgeForClass(
                classLoader, NETWORK_SPEED_VIEW_CLASS);
        if (bridge != null) {
            try {
                java.util.List<ClassData> matches = bridge.findClass(FindClass.create()
                        .searchPackages(SYSTEMUI_PACKAGE)
                        .matcher(ClassMatcher.create()
                                .superClass("android.os.Handler")
                        )
                );
                for (ClassData cd : matches) {
                    String name = cd.getName();
                    if (name.startsWith(NETWORK_SPEED_VIEW_CLASS + "$")) {
                        if (DEBUG) log("DEXKit found Handler inner class: " + name);
                        return classLoader.loadClass(name);
                    }
                }
            } catch (Throwable ignored) {}
        }
        // 回退：遍历常见内部类索引
        for (int i = 1; i <= 10; i++) {
            try {
                Class<?> cls = classLoader.loadClass(NETWORK_SPEED_VIEW_CLASS + "$" + i);
                // 验证是 Handler 子类（有 handleMessage 方法）
                try {
                    cls.getDeclaredMethod("handleMessage", android.os.Message.class);
                    if (DEBUG) log("Found Handler inner class at index " + i);
                    return cls;
                } catch (NoSuchMethodException ignored) {}
            } catch (ClassNotFoundException ignored) {}
        }
        throw new RuntimeException("Cannot find NetworkSpeedView Handler inner class");
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/statusbar/SystemUINetworkSpeeddoublelayerHook.kt
package com.qimian233.ztool.hook.modules.systemui.statusbar

import android.annotation.SuppressLint
import android.content.Context
import android.net.TrafficStats
import android.os.Message
import android.text.Html
import android.util.AttributeSet
import android.util.TypedValue
import com.qimian233.ztool.data.keys.PreferenceKeys
import com.qimian233.ztool.data.keys.ScopeKeys
import com.qimian233.ztool.dexindex.base.DexIndexConstants
import com.qimian233.ztool.hook.base.AppHookModule
import com.qimian233.ztool.hook.base.DexIndexStore
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam
import java.lang.reflect.Constructor
import java.lang.reflect.Field
import java.lang.reflect.Method
import java.text.DecimalFormat
import java.util.Locale
import java.util.WeakHashMap

/**
 * SystemUI network speed display hook module.
 * Function: shows real-time uplink/downlink network speed in the status bar; supports
 * custom text size and display format.
 */
@SuppressLint("PrivateApi")
class SystemUINetworkSpeeddoublelayerHook : AppHookModule() {

    // Store the last traffic data per instance
    private val lastRxBytesMap = WeakHashMap<Any, Long>()
    private val lastTxBytesMap = WeakHashMap<Any, Long>()
    private val lastUpdateTimeMap = WeakHashMap<Any, Long>()

    override fun getModuleName(): String = PreferenceKeys.SYSTEMUI_NETWORK_SPEED_DOUBLELAYER.name

    override fun getTargetPackages(): Array<String> = arrayOf(SYSTEMUI_PACKAGE)

    override fun handleLoadPackage(param: PackageLoadedParam) {
        val classLoader = param.defaultClassLoader
        val packageName = param.packageName
        if (SYSTEMUI_PACKAGE == packageName) {
            hookSystemUINetworkSpeed(classLoader)
        }
    }

    private fun hookSystemUINetworkSpeed(classLoader: ClassLoader) {
        try {
            logger.info("Starting to hook SystemUI NetworkSpeedView")

            // Hook the NetworkSpeedView constructor
            val ctor: Constructor<*> = classLoader.loadClass(NETWORK_SPEED_VIEW_CLASS)
                .getDeclaredConstructor(
                    Context::class.java,
                    AttributeSet::class.java,
                    Int::class.javaPrimitiveType
                )
            hookWithId(ctor, "ctor") { chain ->
                chain.proceed()
                initNetworkSpeedView(chain.thisObject)
                null
            }

            // Hook the Handler's handleMessage method
            hookNetworkSpeedHandler(classLoader)

            logger.info("Successfully hooked NetworkSpeedView")
        } catch (t: Throwable) {
            logger.error("Error hooking NetworkSpeedView", t)
        }
    }

    private fun initNetworkSpeedView(networkSpeedView: Any) {
        try {
            val cl = networkSpeedView.javaClass
            // Get the initial traffic data
            val initialRxBytes = getTotalRxBytes()
            val initialTxBytes = getTotalTxBytes()

            // Store the initial data
            lastRxBytesMap[networkSpeedView] = initialRxBytes
            lastTxBytesMap[networkSpeedView] = initialTxBytes
            lastUpdateTimeMap[networkSpeedView] = System.currentTimeMillis()

            // Adjust the text size
            try {
                // Get the current text size and increase it
                val getTextSizeMethod: Method = findMethod(cl, "getTextSize")
                val textSizeResult = getTextSizeMethod.invoke(networkSpeedView)
                val currentTextSize = textSizeResult as? Float ?: 8.0f
                val newTextSize = currentTextSize * 1.1f // +10%

                val setTextSizeMethod: Method =
                    findMethod(cl, "setTextSize", Int::class.javaPrimitiveType, Float::class.javaPrimitiveType)
                setTextSizeMethod.invoke(
                    networkSpeedView, TypedValue.COMPLEX_UNIT_PX, newTextSize
                )

                logger.debug("Adjusted text size from $currentTextSize to $newTextSize")
            } catch (sizeError: Throwable) {
                logger.error("Error adjusting text size", sizeError)
            }

            logger.debug("Initialized NetworkSpeedView instance")
        } catch (t: Throwable) {
            logger.error("Error initializing NetworkSpeedView", t)
        }
    }

    private fun hookNetworkSpeedHandler(classLoader: ClassLoader) {
        try {
            // Find NetworkSpeedView's inner Handler class via DEXKit (replacing hardcoded $3)
            val handlerClass = findHandlerInnerClass(classLoader)
            val handleMessageMethod: Method =
                handlerClass.getDeclaredMethod("handleMessage", Message::class.java)
            hookWithId(handleMessageMethod, "handle_message") { chain ->
                val handler = chain.thisObject
                val handlerCls = handler.javaClass
                val this0Field: Field = handlerCls.getDeclaredField("this\$0")
                this0Field.isAccessible = true
                val networkSpeedView = this0Field.get(handler)
                    ?: return@hookWithId chain.proceed()

                // Get the message object
                val message = chain.args[0]
                val what = message.javaClass.getDeclaredField("what").getInt(message)

                if (what == 10) { // speed update message
                    handleSpeedUpdate(networkSpeedView, handler)
                    return@hookWithId null // block the original handling
                } else if (what == 1) { // format/display message
                    handleSpeedDisplay(networkSpeedView, message)
                    return@hookWithId null // block the original handling
                }
                chain.proceed()
            }
        } catch (t: Throwable) {
            logger.error("Error hooking NetworkSpeed handler", t)
        }
    }

    private fun handleSpeedUpdate(networkSpeedView: Any, handler: Any) {
        try {
            val handlerCls = handler.javaClass
            // Remove the previous message
            findMethod(handlerCls, "removeMessages", Int::class.javaPrimitiveType).invoke(handler, 10)

            // Check whether the network speed should be shown
            val isIconVisibleResult = networkSpeedView.javaClass
                .getDeclaredMethod("isIconVisible").invoke(networkSpeedView)
            val shouldShow = java.lang.Boolean.TRUE == isIconVisibleResult

            if (!shouldShow) {
                return
            }

            // Get the current traffic stats
            val currentRxBytes = getTotalRxBytes()
            val currentTxBytes = getTotalTxBytes()
            val currentTime = System.currentTimeMillis()

            // Get the previous data
            val lastRxBytes = lastRxBytesMap[networkSpeedView]
            val lastTxBytes = lastTxBytesMap[networkSpeedView]
            val lastUpdateTime = lastUpdateTimeMap[networkSpeedView]

            if (lastRxBytes == null || lastTxBytes == null || lastUpdateTime == null) {
                // First update, record data without computing speed
                lastRxBytesMap[networkSpeedView] = currentRxBytes
                lastTxBytesMap[networkSpeedView] = currentTxBytes
                lastUpdateTimeMap[networkSpeedView] = currentTime
            } else {
                // Compute the time difference (seconds)
                val timeDiff = (currentTime - lastUpdateTime) / 1000
                if (timeDiff > 0) {
                    // Compute uplink/downlink speed (bytes/second)
                    val downSpeed = (currentRxBytes - lastRxBytes) / timeDiff
                    val upSpeed = (currentTxBytes - lastTxBytes) / timeDiff

                    // Log debug info
                    logger.debug(
                        String.format(
                            Locale.US,
                            "Successfully updated speed - downSpeed=%d, upSpeed=%d, timeDiff=%d",
                            downSpeed, upSpeed, timeDiff
                        )
                    )

                    // Send the display message
                    val message = findMethod(handlerCls, "obtainMessage").invoke(handler)
                    if (message != null) {
                        val msgCls = message.javaClass
                        msgCls.getDeclaredField("what").setInt(message, 1)
                        msgCls.getDeclaredField("obj").set(message, longArrayOf(downSpeed, upSpeed))
                        findMethod(handlerCls, "sendMessage", Message::class.java)
                            .invoke(handler, message)
                    }

                    // Update the data
                    lastRxBytesMap[networkSpeedView] = currentRxBytes
                    lastTxBytesMap[networkSpeedView] = currentTxBytes
                    lastUpdateTimeMap[networkSpeedView] = currentTime
                }
            }

            val refreshInterval = (xposed.getRemotePreferences(PREFS_NAME)
                .getFloat("systemui_network_speed_refresh_interval", 3.0f) * 1000.0).toLong()
            // Schedule the next update
            findMethod(handlerCls, "sendEmptyMessageDelayed", Int::class.javaPrimitiveType, Long::class.javaPrimitiveType)
                .invoke(handler, 10, refreshInterval)
        } catch (t: Throwable) {
            logger.error("Error in speed update", t)
        }
    }

    private fun handleSpeedDisplay(networkSpeedView: Any, message: Any) {
        try {
            val speeds = message.javaClass.getDeclaredField("obj").get(message) as? LongArray
                ?: return
            if (speeds.size != 2) return

            val downSpeed = speeds[0]
            val upSpeed = speeds[1]

            // Format uplink/downlink speed
            val downText = formatSpeed(downSpeed)
            val upText = formatSpeed(upSpeed)

            // Create the double-layer display text with HTML formatting, adjusting font size
            val displayText = "<font size='5'><b>▴ " + upText + "</b></font><br/>" +
                "<font size='5'><b>▾ " + downText + "</b></font>"

            // Set the text using HTML formatting
            findMethod(networkSpeedView.javaClass, "setText", CharSequence::class.java)
                .invoke(networkSpeedView, Html.fromHtml(displayText, Html.FROM_HTML_MODE_LEGACY))
        } catch (t: Throwable) {
            logger.error("Error in speed display", t)
        }
    }

    private fun formatSpeed(speedBytes: Long): String {
        if (speedBytes <= 0) {
            return "0.00 B/s"
        }

        val speed: Double
        val unit: String

        if (speedBytes >= 1073741824) { // 1 GB
            speed = speedBytes / 1073741824.0
            unit = "G/s"
        } else if (speedBytes >= 1048576) { // 1 MB
            speed = speedBytes / 1048576.0
            unit = "M/s"
        } else if (speedBytes >= 1024) { // 1 KB
            speed = speedBytes / 1024.0
            unit = "K/s"
        } else {
            speed = speedBytes.toDouble()
            unit = "B/s"
        }

        // Choose precision based on the speed value
        val formatPattern = if (speed >= 100) {
            "0"
        } else if (speed >= 10) {
            "0.0"
        } else {
            "0.00"
        }

        val df = DecimalFormat(formatPattern)
        return df.format(speed) + " " + unit
    }

    private fun getTotalRxBytes(): Long {
        return try {
            val result = TrafficStats::class.java
                .getDeclaredMethod("getTotalRxBytes").invoke(null)
            result as? Long ?: 0L
        } catch (t: Throwable) {
            logger.error("Error getting Rx bytes", t)
            0
        }
    }

    private fun getTotalTxBytes(): Long {
        return try {
            val result = TrafficStats::class.java
                .getDeclaredMethod("getTotalTxBytes").invoke(null)
            result as? Long ?: 0L
        } catch (t: Throwable) {
            logger.error("Error getting Tx bytes", t)
            0
        }
    }

    /**
     * Find NetworkSpeedView's inner Handler subclass via reflection.
     * Iterates possible inner class indexes, replacing the hardcoded $3.
     */
    private fun findHandlerInnerClass(classLoader: ClassLoader): Class<*> {
        // First try the Handler subclass from the offline index
        val indexed = DexIndexStore.string(
            xposed, SYSTEMUI_PACKAGE,
            DexIndexConstants.ModuleKeys.SYSTEMUI_NETWORK_SPEED_DOUBLELAYER,
            DexIndexConstants.Keys.HANDLER_INNER_CLASS
        )
        if (indexed != null) {
            try {
                val cls = classLoader.loadClass(indexed)
                logger.debug("Loaded Handler inner class from dex index: $indexed")
                return cls
            } catch (_: ClassNotFoundException) {
            }
        }
        // Fallback: iterate common inner class indexes
        for (i in 1..10) {
            try {
                val cls = classLoader.loadClass(NETWORK_SPEED_VIEW_CLASS + "$" + i)
                // Verify it is a Handler subclass (has the handleMessage method)
                try {
                    cls.getDeclaredMethod("handleMessage", Message::class.java)
                    logger.debug("Found Handler inner class at index $i")
                    return cls
                } catch (_: NoSuchMethodException) {
                }
            } catch (_: ClassNotFoundException) {
            }
        }
        throw RuntimeException("Cannot find NetworkSpeedView Handler inner class")
    }

    companion object {
        private const val PREFS_NAME = "xposed_module_config"
        private val SYSTEMUI_PACKAGE = ScopeKeys.SYSTEM_UI.packageName
        private const val NETWORK_SPEED_VIEW_CLASS = "com.android.systemui.zui.NetworkSpeedView"
    }
}
